File: core/src/main/java/com/pyreon/systems/SpriteBatchDrawSystem.java
package com.pyreon.systems;

import com.artemis.Aspect;
import com.artemis.BaseEntitySystem;
import com.artemis.ComponentMapper;
import com.artemis.utils.IntBag;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.pyreon.components.PositionComponent;
import com.pyreon.components.SpriteComponent;
import com.pyreon.components.TextureComponent;
import com.pyreon.tournament.Tournament;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class SpriteBatchDrawSystem extends BaseEntitySystem {
    private static Logger log = LogManager.getLogger(SpriteBatchDrawSystem.class.getSimpleName());

    ComponentMapper<PositionComponent> pm;
    ComponentMapper<SpriteComponent> sm;
    ComponentMapper<TextureComponent> tm;
    Tournament game;

    public SpriteBatchDrawSystem(Tournament game) {
        super(Aspect.all(PositionComponent.class).one(SpriteComponent.class, TextureComponent.class));
        this.game = game;
    }

    @Override
    protected void processSystem() {
        IntBag actives = subscription.getEntities();
        int[] ids = actives.getData();

        game.batch.begin();
        for (int i = 0, s = actives.size(); i < s; i++) {
            PositionComponent position = pm.get(ids[i]);
            if (sm.has(ids[i])) {
                Sprite sprite = sm.get(ids[i]).sprite;
                sprite.setPosition(position.x - sprite.getWidth()/2, position.y - sprite.getHeight()/2);
                sprite.draw(game.batch);
            } else if (tm.has(ids[i])) {
                game.batch.draw(tm.get(ids[i]).texture, position.x, position.y);
            }
        }
        game.batch.end();
    }
}

File: core/src/main/java/com/pyreon/systems/SpriteBatchDrawSystem.kt
package com.pyreon.systems

import com.artemis.Aspect
import com.artemis.BaseEntitySystem
import com.artemis.ComponentMapper
import com.artemis.systems.IteratingSystem
import com.artemis.utils.IntBag
import com.badlogic.gdx.graphics.g2d.Sprite
import com.pyreon.components.PositionComponent
import com.pyreon.components.SpriteComponent
import com.pyreon.components.TextureComponent
import com.pyreon.tournament.Tournament
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

class SpriteBatchDrawSystem(internal var game: Tournament) : IteratingSystem(Aspect.all(PositionComponent::class.java).one(SpriteComponent::class.java, TextureComponent::class.java)) {

    internal var pm: ComponentMapper<PositionComponent>? = null
    internal var sm: ComponentMapper<SpriteComponent>? = null
    internal var tm: ComponentMapper<TextureComponent>? = null

    override fun begin(){
        game.batch.begin()
    }
    override fun process(entityId: Int) {
        val position = pm!!.get(entityId)
        if (sm!!.has(entityId)) {
            val sprite = sm!!.get(entityId).sprite
            sprite!!.setPosition(position.x - sprite.width / 2, position.y - sprite.height / 2)
            sprite.draw(game.batch)
        } else if (tm!!.has(entityId)) {
            game.batch.draw(tm!!.get(entityId).texture, position.x, position.y)
        }
    }

    override fun end(){
        game.batch.end()
    }

    companion object {
        private val log = LogManager.getLogger(SpriteBatchDrawSystem::class.java.simpleName)
    }
}
